packages/network/src/logic/proxy/ProxyStreamConnectionClient.ts

```diff
@@ -0,0 +1,289 @@
+            .splice(0, connectionCount)
+        await Promise.allSettled(proxiesToDisconnect.map((node) => this.closeConnection(streamPartId, node)))
+    }
+
+    private async closeConnection(streamPartId: StreamPartID, targetNodeId: NodeId): Promise<void> {
+        if (this.getConnections(streamPartId).has(targetNodeId)
+            && this.streamPartManager.hasOnewayConnection(streamPartId, targetNodeId)
+        ) {
+            logger.info(`Close proxy connection to ${targetNodeId} on ${streamPartId}`)
+            await this.nodeToNode.leaveStreamOnNode(targetNodeId, streamPartId)
+            this.node.emit(NodeEvent.ONE_WAY_CONNECTION_CLOSED, targetNodeId, streamPartId)
+            this.removeConnection(streamPartId, targetNodeId)
+        }
+    }
+
+    private getConnections(streamPartId: StreamPartID): Map<NodeId, ProxyDirection> {
+        return this.connections.get(streamPartId) ?? new Map()
+    }
+
+    private hasConnection(nodeId: NodeId, streamPartId: StreamPartID): boolean {
+        if (!this.definitions.has(streamPartId)) {
+            return false
+        }
+        return this.getConnections(streamPartId).has(nodeId)
+    }
+
+    private removeConnection(streamPartId: StreamPartID, nodeId: NodeId): void {
+        if (this.connections.has(streamPartId)) {
+            this.connections.get(streamPartId)!.delete(nodeId)
+        }
+        this.streamPartManager.removeNodeFromStreamPart(streamPartId, nodeId)
+    }
+
+    private processHandshakeResponse(message: ProxyConnectionResponse, nodeId: NodeId): void {
+        const streamPartId = message.getStreamPartID()
+        if (message.accepted) {
+            if (message.direction === ProxyDirection.PUBLISH) {
+                this.streamPartManager.addOutOnlyNeighbor(streamPartId, nodeId)
+                this.propagation.onNeighborJoined(nodeId, streamPartId)
+            } else {
+                this.streamPartManager.addInOnlyNeighbor(streamPartId, nodeId)
+            }
+            this.emit(Event.CONNECTION_ACCEPTED, nodeId, streamPartId, message.direction)
+        } else {
+            this.removeConnection(streamPartId, nodeId)
+            this.emit(
+                Event.CONNECTION_REJECTED,
+                nodeId,
+                streamPartId,
+                message.direction,
+                `Target node ${nodeId} rejected proxy ${message.direction} stream connection ${streamPartId}`
+            )
+        }
+    }
+
+    async onNodeDisconnected(streamPartId: StreamPartID, nodeId: NodeId): Promise<void> {
+        this.removeConnection(streamPartId, nodeId)
+        await retry(() => this.updateConnections(streamPartId), 'Updating proxy connections', this.abortController.signal)
+    }
+
+    isProxiedStreamPart(streamPartId: StreamPartID, direction: ProxyDirection): boolean {
+        if (this.definitions.get(streamPartId) && this.getConnections(streamPartId).size > 0) {
```

Comment:
This method answers the question whether some `streamPart` is proxied or not? If we called setProxies for the streamPart but have not yet formed a connections (or the connections are lost and reconnect again), this should return `true`? I.e. should we remove the `this.getConnections(streamPartId).size > 0` check?